### ISSUE TITLE
Incorrect "Signature incompatible with supertype" errors when using an ABC

### ISSUE THREAD

**Issue by andlaus** [python/mypy#12758]
**Bug Report**

when checking a class derived from an abstract base class using mypy 0.950, I get incorrect "signature is incompatible with supertype" errors. These errors only seem to occur if the method in the base class is decorated with the `@abstractclassmethod` decorator and even then only if a return type is specified.

**To Reproduce**

Run mypy on the following snippet:

```python
import abc

class A:
    @abc.abstractclassmethod
    def foo(self) -> int:
        return 1

class B(A):
    def foo(self) -> int:
        return 2
```

if either `@abc.abstractclassmethod` or the return type specifier of the method is removed, mypy is successful.

**Expected Behavior**

```
$ mypy bla.py 
Success: no issues found in 1 source file
```

**Actual Behavior**

```
$ mypy bla.py 
bla.py:9: error: Signature of "foo" incompatible with supertype "A"
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 0.950
- Python version used: 3.8.13
- Operating system and version: Opensuse Tumbleweed

**Comment by JelleZijlstra** [comment]
Arguably a duplicate of #6244, though you're reporting a different symptom.

**Comment by andlaus** [comment]
I agree that it might be a duplicate but I'm also not sure. it seems like mypy has an issue with the `@abstractclassmethod` decorator, but the issue seems to have cropped up recently: mypy 0.942  (my previously used version), didn't have this issue...

**Comment by andlaus** [comment]
Changing the example above not to use deprecated constructs:
```python
import abc

class A:
    @classmethod
    @abc.abstractmethod
    def foo(self) -> int:
        return 1

class B(A):
    def foo(self) -> int:
        return 2
```
also does not help, although the error message gets slightly more verbose:
```
$ mypy bla.py 
bla.py:10: error: Signature of "foo" incompatible with supertype "A"
bla.py:10: note:      Superclass:
bla.py:10: note:          @classmethod
bla.py:10: note:          def foo(self) -> int
bla.py:10: note:      Subclass:
bla.py:10: note:          def foo(self) -> int
Found 1 error in 1 file (checked 1 source file)
```

**Comment by andlaus** [comment]
next update: it works if `@classmethod` is added to the `foo()` method of `B`, i.e. a method with the `@classmethod` decorator in the base class also needs to be declared with this decorator in all derived classes (but _not_ vice-versa). To me this behaviour seems to be (slightly) incorrect?

**Comment by dolfinus** [comment]
`classmethod` is not the same as a regular method - for the `classmethod` first method argument is `cls`, not `self`. It does not look like an issue

**Comment by yawhide** [comment]
I tihnk this is a related issue. Y'all can let me know if not.

I have this example:

```
from abc import ABC, abstractclassmethod
from typing import Any


class MyBaseClass(ABC):
    @classmethod
    @abstractclassmethod
    def dosomething(cls, name: str) -> Any:
        return name


class MyChildClass(MyBaseClass):
    @classmethod
    def dosomething(cls, name: str) -> Any:
        return name

```

The error I get:
```
test.py:14: error: Signature of "dosomething" incompatible with supertype "MyBaseClass"  [override]
```

**Comment by erictraut** [comment]
`abstractclassmethod` has been [deprecated since Python 3.3](https://docs.python.org/3/library/abc.html#abc.abstractclassmethod). If you switch to the recommended alternative (use `@classmethod` with `@abstractmethod`), it will work fine at runtime and mypy will be happy with it.
